launcher/minecraft/mod/tasks/LocalModParseTask.cpp

```diff
@@ -401,6 +428,36 @@ bool processZIP(Mod& mod, ProcessingLevel level)
 
         mod.setDetails(details);
         return true;
+    } else if (zip.setCurrentFile("META-INF/nil/mappings.json")) {
+        // nilloader uses the filename of the metadata file for the modid, so we can't know the exact filename
+        // thankfully, there is a good file to use as a canary so we don't look for nil meta all the time
+
+        QStringList foundNilMetas;
+        for (auto& fname : zip.getFileNameList()) {
+            if (fname.endsWith(".nilmod.css")) {
+                foundNilMetas.append(fname);
+            }
+        }
+
+        if (foundNilMetas.size() > 1 && foundNilMetas.at(0) == "nilloader.nilmod.css") {
+            // nilmods can shade nilloader to be able to run as a standalone agent - which includes nilloader's own meta file
+            // so if there is more than one meta file, ignore nilloader's meta, since it's not the actual mod
+            foundNilMetas.removeFirst();
+        }
+
+        if (zip.setCurrentFile(foundNilMetas.at(0))) {
```

Comment:
Is it guaranteed that the *main* metadata file is also the first one we find?

Reply:
I don't believe there'd be a good way to handle a jar with multiple nilmods in this UI — it's *technically possible*, and NilLoader will load them all, sharing the jar, but no one has yet done this and it's kind of an edge case.

The real reason nilmod CSS files have the modid in them is to avoid conflicts when classpaths get merged, and to avoid conflicts when nilmods shade NilLoader themselves.

Reply:
Yeah, this is essentially the same assumption as the first mod entry we find in `mods.toml` being the main mod - someone could put a submod first for whatever reason and that's the mod we would get metadata for.